code/modules/aquarium/aquarium.dm

```diff
@@ -0,0 +1,271 @@
+	 * min_offset + max_offset = this value is returned for top layer mode
+	 * min_offset + max_offset + 1 = this is used for glass overlay
+	 */
+	//optional todo: hook up sending surface changed on aquarium changing layers
+	switch(layer_type)
+		if(AQUARIUM_LAYER_MODE_BOTTOM)
+			return layer + AQUARIUM_MIN_OFFSET
+		if(AQUARIUM_LAYER_MODE_TOP)
+			return layer + AQUARIUM_MAX_OFFSET
+		if(AQUARIUM_LAYER_MODE_AUTO)
+			var/chosen_layer = layer + AQUARIUM_MIN_OFFSET + AQUARIUM_LAYER_STEP
+			while((chosen_layer in used_layers) && (chosen_layer <= layer + AQUARIUM_MAX_OFFSET))
+				chosen_layer += AQUARIUM_LAYER_STEP
+			used_layers += chosen_layer
+			return chosen_layer
+
+/obj/structure/aquarium/proc/free_layer(value)
+	used_layers -= value
+
+/obj/structure/aquarium/proc/get_surface_properties()
+	. = list()
+	.[AQUARIUM_PROPERTIES_PX_MIN] = aquarium_zone_min_px
+	.[AQUARIUM_PROPERTIES_PX_MAX] = aquarium_zone_max_px
+	.[AQUARIUM_PROPERTIES_PY_MIN] = aquarium_zone_min_py
+	.[AQUARIUM_PROPERTIES_PY_MAX] = aquarium_zone_max_py
+
+/obj/structure/aquarium/update_overlays()
+	. = ..()
+	if(panel_open)
+		. += "panel"
+
+	//Glass overlay goes on top of everything else.
+	var/mutable_appearance/glass_overlay = mutable_appearance(icon,broken ? broken_glass_icon_state : glass_icon_state,layer=AQUARIUM_MAX_OFFSET-1)
+	. += glass_overlay
+
+/obj/structure/aquarium/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>Alt-click to [panel_open ? "close" : "open"] the control panel.</span>"
+
+/obj/structure/aquarium/AltClick(mob/user)
+	if(!user.canUseTopic(src, BE_CLOSE))
+		return ..()
+	panel_open = !panel_open
+	update_icon()
+
+/obj/structure/aquarium/wrench_act(mob/living/user, obj/item/I)
+	if(default_unfasten_wrench(user,I))
+		return TRUE
+
+/obj/structure/aquarium/attackby(obj/item/I, mob/living/user, params)
+	if(!broken)
+		// This signal exists so we common items instead of adding component on init can just register creation of one in response.
+		// This way we can avoid the cost of 9999 aquarium components on rocks that will never see water in their life.
+		SEND_SIGNAL(I,COMSIG_AQUARIUM_BEFORE_INSERT_CHECK,src)
+		if(SEND_SIGNAL(I,COMSIG_AQUARIUM_INSERT_READY,src) & AQUARIUM_CONTENT_READY_TO_INSERT)
+			if(user.transferItemToLoc(I,src))
+				SEND_SIGNAL(I,COMSIG_AQUARIUM_INSERTED,src)
+				update_icon()
+				return TRUE
+		else if(istype(I,/obj/item/fish_feed))
+			feed_fish(user,I)
+			return TRUE
+		else
+			return ..()
+	else
+		var/obj/item/stack/sheet/glass/G = I
+		if(istype(G))
+			if(G.get_amount() < 2)
+				to_chat(user, "<span class='warning'>You need two glass sheets to fix the case!</span>")
+				return
+			to_chat(user, "<span class='notice'>You start fixing [src]...</span>")
+			if(do_after(user, 20, target = src))
+				G.use(2)
+				broken = FALSE
+				obj_integrity = max_integrity
+				update_icon()
+			return TRUE
+	return ..()
+
+/obj/structure/aquarium/attacked_by(obj/item/I, mob/living/user)
```

Comment:
A lot of this should be in the components themselves by just listening for the attacked_by signal